File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/NetWork/NetCmdDispatch/NetCmdDispatchHandle.cpp
#include "MyProject.h"
#include "NetCmdDispatchHandle.h"
#include "Prequisites.h"
#include "UtilStr.h"
#include "MByteBuffer.h"
#include "AddOnceEventDispatch.h"
#include "NetCmdDispatchInfo.h"
#include "MClassInfo.h"
#include "MClassMacros.h"
#include "MClassFactory.h"
#include "UtilConvert.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(NetCmdDispatchHandle, GObject)

NetCmdDispatchHandle::NetCmdDispatchHandle()
{
	
}

NetCmdDispatchHandle::~NetCmdDispatchHandle()
{
	
}

void NetCmdDispatchHandle::init()
{

}

void NetCmdDispatchHandle::dispose()
{

}

void NetCmdDispatchHandle::addParamHandle(uint8 paramId, EventDispatchDelegate handle)
{
	if (!this->mId2HandleDic.containsKey(paramId))
	{
		AddOnceEventDispatch* addOnceEventDispatch = MY_NEW AddOnceEventDispatch();
		this->mId2HandleDic.add(paramId, addOnceEventDispatch);
	}
	else
	{
		// 日志
	}

	this->mId2HandleDic.value(paramId)->addEventHandle(handle);
}

void NetCmdDispatchHandle::removeParamHandle(uint8 paramId, EventDispatchDelegate handle)
{
	if (this->mId2HandleDic.containsKey(paramId))
	{
		this->mId2HandleDic.value(paramId)->removeEventHandle(handle);
	}
	else
	{

	}
}

void NetCmdDispatchHandle::handleMsg(uint eventId, IDispatchObject* dispObj)
{
	NetCmdDispatchInfo* cmd = (NetCmdDispatchInfo*)dispObj;

	if (this->mId2HandleDic.containsKey(cmd->byParam))
	{
		this->mId2HandleDic[cmd->byParam]->dispatchEvent(cmd->bu);
	}
	else
	{
		GLogSys->log(
			UtilStr::Format(
				"Msg Not Handle: byCmd = {0}, byParam = {1}", 
				UtilConvert::convInt2Str(cmd->byCmd),
				UtilConvert::convInt2Str(cmd->byParam)
			)
		);
	}
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Network/CmdDispatch/NetCmdDispatchHandle.lua
local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "NetCmdDispatchHandle";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mId2HandleDic = GlobalNS.new(GlobalNS.MDictionary);
end

function M:addParamHandle(paramId, pThis, func)
    if(not self.mId2HandleDic:containsKey(paramId)) then
        local disp = GlobalNS.new(GlobalNS.AddOnceEventDispatch);
        self.mId2HandleDic:add(paramId, disp);
    else
        GCtx.mLogSys:log("Msg Id Already Register paramId = " .. paramId, GlobalNS.LogTypeId.eLogCommon);
    end

    self.mId2HandleDic:value(paramId):addEventHandle(pThis, func, 0);
end

function M:removeParamHandle(paramId, pThis, func)
    if(self.mId2HandleDic.ContainsKey(paramId)) then
        self.mId2HandleDic:value(paramId):removeEventHandle(pThis, func, 0);
    else
        
    end
end

function M:handleMsg(dispObj)
	local cmd = dispObj;
    GCtx.mLogSys:log("NetCmdDispatchHandle Start handleMsg", GlobalNS.LogTypeId.eLogCommon);
	
    if(self.mId2HandleDic:containsKey(cmd.byParam)) then
        GCtx.mLogSys:log("NetCmdDispatchHandle In handleMsg", GlobalNS.LogTypeId.eLogCommon);
        self.mId2HandleDic:value(cmd.byParam):dispatchEvent(cmd.bu);
    else
        
    end
end

return M;